### ISSUE TITLE
UAA not working with eureka disabled

### ISSUE THREAD

**Issue by xetys** [jhipster/generator-jhipster#9027]
##### **Overview of the issue**

consider the following JDL:

```
application {
	config {
  		baseName UserService
        	applicationType uaa
        	cacheProvider no
        	enableHibernateCache no
        	buildTool gradle
        	serviceDiscoveryType no
    }
}

application {
	config {
    		baseName GateWay
        	applicationType gateway
        	authenticationType uaa
        	uaaBaseName "UserService"
        	devDatabaseType no
        	prodDatabaseType no
        	databaseType no
        	buildTool gradle
        	serviceDiscoveryType no
    }
}

deployment {
  deploymentType docker-compose
  appsFolders [UserService, GateWay]
  dockerRepositoryName "xetys"
}

```

If we generate this at it is, the setup won't work as of following reasons:

1. we cannot reach any microservices using routes over the gateway as both zuul is missing and no proxy or what ever is present (see #8967)
2. if we enable zuul (what we can do without any additions to dependencies in gradle, as it is already there, while it lacks using maven), then we need to add static routes manually. But here, we need to respect the context-path @saturnism invented in #7697.
3. if we respect the context-path in the zuul routes, we still have problems, as the oauth2 endpoints are pointing to "userservice" instead of "userservice/UserService"
4. if we fix that, the `UaaTokenEndpointClient` and `UaaSignatureVerifierClient` will try to use the `loadBalancedRestTemplate`, which won't work if eureka is not active. We should use the `vanillaRestTemplate` instead
5. without pointing out the ports in the UAA/OAuth2 related endpoints, it will assume using port 80, which is usually not the correct one
6. if we even fix the ports, the gateway won't resolve the correct hostnames, as our docker-compose generate names them like `userservice-app` but not `userservice`

I described the behavior with docker-compose by example, but it will happen the same way using kubernetes.

On the other hand, if we re-activate eureka, everything works like intended.

This issue is very much like #8967, but this one is just an improvment request, while this is a critical issue and should be fixed ASAP.

##### **Suggest a Fix**

Out of my 6 points mentioned above, I see the following TODOs:

- [x] change docker-compose service name convention
- [x] remove the context-path extension invented by @saturnism. Regarding this, I would like to do this in general. I was thinking about what you wanted to do with ingress and stuff, but I suggest to remove that completely for microserivces, as
  - it doesn't work with UAA it all
  - we cannot use stuff like ratelimiter or user defined filter beans for normal microservices in general
I would prefer to use static zuul routes for non-eureka microservices
- [ ] introduce static zuul routes for non-eureka microservices
- [x] use `vanillaRestTemplate` for the refresh-token grant logic

This would both fixing UAA in non-eureka microservices AND improve the docker-compose and kubernetes experience for non-eureka microservice in general, as we then could use both for development and production.

I would like to priorize this TODOs ASAP, in case @saturnism @PierreBesson @ruddell don't have any concerns on my plan. This could possibly affect GAE, but I'm not sure.

WDYT?

##### **JHipster Version(s)**
master

-   [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by PierreBesson** [comment]
@xetys Apart from the general discussion on were we want to go with the non Eureka/Spring Cloud service discovery option. @xetys do you have the bandwith to work on this in the coming months ? We have made little progress on this because it implies deep architecture changes and several weeks of work to do it properly.

Maybe a problem with our approach has been to prioritise Istio before solving the story with no discovery microservices on docker-compose/kubernetes.

**Comment by xetys** [comment]
Oh, I don't see that it is so much work actually. I did the steps above in like an hour and it would take me a few hours to do that in the generator. It doesn't harm the Istio stuff too much. It would just reduce our usage on ingress like we do currently. But as of the points mentioned above, I'm not a fan of routing the microservices using Ingress and bypassing the gateway, as we do a lot of stuff there, like UAA cookies, ratelimiting, access control etc.

**Comment by pascalgrimaud** [comment]
@xetys : it is opened since 5 months. Do you plan to work on this ?

**Comment by xetys** [comment]
Yes, actually I do....stuff with bachelor and family was moving that plan away...I'm trying to get into that asap

**Comment by pascalgrimaud** [comment]
no worry no stress @xetys, it's just a reminder
Family is more important than anything else.

**Comment by pascalgrimaud** [comment]
Then, it has been opened since 9 months now.
So let's close this as no one seems interested to work on this and nothing has been done since.

**Comment by Pranavism** [comment]
this is not getting fixed anytime soon ? I was wondering if a fix can be provided then UAA based architecture can be deployed with istio on k8 instead of dependency over eureka .

**Comment by MathieuAA** [comment]
@Pranavism you can always contribute and fix this if you want to. We'd be happy to help you if need be

**Comment by xetys** [comment]
I'm very sorry guys, I had my reasons for inactivity.

I'm reopening this issue as I started working on it

**Comment by PierreBesson** [comment]
@xetys Sorry if what I say doesn't make sense as I don't know much in the workings of Oauth2 but isn't there a way to make the UAA behave like a standard Oauth2 server and reuse the code of the JHipster oauth2 option as much as possible. 
Ideally, we would prefer to have only 2 options to choose from for JHipster Authentication: JWT and Oauth2 without sacrificing the UAA option as maintaining end 2 end tests for 3 different options is very hard for @pascalgrimaud and the team.

**Comment by xetys** [comment]
For this I should do some wider experiments with oauth2 and see, what we can summarize. But anyways, I'd love to see uaa back working in the most of our options.

I guess I'll have more time from now to support this on the long term

**Comment by PierreBesson** [comment]
@xetys We have about 150 UAA generated every day in our stats vs 1500 microservices (and the stats are opt-in).  This means that the UAA option is still popular but we have few contributors, it doesn't have someone like @mraible to maintain support. It would be for the best if this can be supported client-side using common code with the oauth2 support.

![Screenshot from 2019-11-13 17-43-35](https://user-images.githubusercontent.com/513471/68784708-42b2b080-063d-11ea-878a-621637858c96.png)

**Comment by xetys** [comment]
Well, I was quite inactive in the last months, so I couldn't support it. But stuff is changing in my carrier and it could open a lot of work- and spare time on contribution again, especially in the UAA case.

In other words, it's likely that I will have time to support UAA on a wider level the next time


About the OAuth2 vs. UAA question, I'm still convinced, that these approaches fit for very different needs. OAuth2 mainly relies on an external identity provider, while UAA is its own identity provider. OAuth2 opens more oauth2 interconnectivity options, while UAA allows changing the User Domain, specific OAuth2 option directly without complex configuration. AFAIK, UAA enables full stateless authorization, while the OAuth2 option is stateful, but outsourced so it's kinda stateless for us. Both solutions implement an SSO. For OAuth2, Keycloak or Octa is your single sign-on. In UAA, thanks to #5812, you can use domain cookies for a different type of SSO, where you log in once and all frontends under one domain can handle tokens.

In summary, I would say both solutions should exist, as they are designed for slightly different problems and both are legit IMHO...

I see the problem was my lack of support...I'm working on it :)

**Comment by pascalgrimaud** [comment]
closed by https://github.com/jhipster/generator-jhipster/pull/10763
